keps/sig-auth/20190116-service-account-external-signing.md

```diff
@@ -0,0 +1,152 @@
+
+### Goals
+
+- Support for out-of-process JWT signing
+- Support for listing public verifying keys
+- Preserve existing behavior and performance for keys not read over a socket
+
+### Non-Goals
+
+- Reading TLS serving certificates and key from a socket or reloading of the API server with new cert and key
+- Reading any other certificates from a file
+
+## Proposal
+
+### Preserve existing behavior
+
+The API server flags `--service-account-key-file` and `--service-account-signing-key-file` will continue be used for reading from files. The API server will detect if the provided files are regular files to be read once at process startup, or unix sockets that should be read on each token request using a new API. Legacy service account verification will skip any socket files.
+
+### Updates to API server token generation
+
+As of Kubernetes v1.13.2, the API server uses the functions `JWTTokenGenerator` and `JWTTokenAuthenticator`. New types that implement the `TokenGenerator` interface and support token validation will be added to `k8s.io/kubernetes/pkg/serviceaccount/`.
+
+### New API
+
+I'm proposing creating a new versioned grpc API under `k8s.io/kubernetes/pkg/serviceaccount`. This will be similar to how the KMS envelope encryption has an API at `k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.proto`
+
+```proto
+syntax = "proto3";
+
+package v1beta1;
+
+service KeyService {
+  // Sign an incoming payload
+  rpc SignPayload(SignPayloadRequest) returns (SignPayloadResponse) {}
+  // List supported Algorithms
+  rpc ListAlgorithms(ListAlgorithmsRequest) returns (ListAlgorithmsResponse) {}
+  // Get a specific key by ID, defaulting to the current public key
+  rpc GetPublicKey(GetPublicKeyRequest) returns (GetPublicKeyResponse) {}
+  // List all active public keys
+  rpc ListPublicKeys(ListPublicKeysRequest) returns (ListPublicKeysResponse) {}
+}
+
+message SignPayloadRequest {
+  // payload is the content to be signed
+  bytes payload = 1;
+  // algorithm specifies which algorithm to sign with
+  string algorithm = 2;
+}
+message SignPayloadResponse {
+  // content returns the signed payload
+  bytes content = 1;
+}
+
+message ListAlgorithmsRequest {}
+message ListAlgorithmsResponse {
+  // signature_algorithms returns the supported algorithms
+  repeated string signature_algorithms = 1;
+}
+
+message PublicKey {
+  // public_key is a PEM encoded public key
+  bytes public_key = 1;
+  // certificate is a concatenated list of PEM encoded x509 certificates
+  bytes certificates = 2;
+  // key_id is the key's ID
+  string key_id = 3;
+  // algorithm states the algorithm the key uses
+  string algorithm = 4;
+  // use states what the key is used for
+  string use = 5 ;
```

Comment:
Maybe `active`?

Reply:
Or hoist `active_key_id` into `ListAlgorithmsResponse`.

Reply:
Moving to a field on ListAlgorithmsResponse